bokeh/util/callback_manager.py

```diff
@@ -3,45 +3,49 @@
 interfaces to classes.
 '''
 from __future__ import absolute_import
+
 from functools import partial
-from inspect import formatargspec, getargspec, isfunction, ismethod
-from types import FunctionType
+from inspect import isfunction, ismethod
+
 from ..events import Event
+from ..util.future import get_param_info, format_signature, signature
 
-def _callback_argspec(callback):
-    '''Bokeh-internal function to get argspec for a callable'''
+# XXX: This can probably be dispensed with if/when python 2 support is dropped
+def _callback_signature(callback):
+    '''Bokeh-internal function to get signature for a callable'''
     if not callable(callback):
         raise ValueError("Callbacks must be callables")
 
     if isfunction(callback) or ismethod(callback):
-        return getargspec(callback)
+        return signature(callback)
     elif isinstance(callback, partial):
-        return getargspec(callback.func)
+        return signature(callback.func)
     else:
-        return getargspec(callback.__call__)
+        return signature(callback.__call__)
 
 def _check_callback(callback, fargs, what="Callback functions"):
     '''Bokeh-internal function to check callback signature'''
-    argspec = _callback_argspec(callback)
-    formatted_args = formatargspec(*argspec)
-    margs = ('self',) + fargs
+    sig = _callback_signature(callback)
+    formatted_args = format_signature(sig)
     error_msg = what + " must have signature func(%s), got func%s"
-    defaults_length = len(argspec.defaults) if argspec.defaults else 0
 
-    if isinstance(callback, FunctionType):
-        if len(argspec.args) - defaults_length != len(fargs):
-            raise ValueError(error_msg % (", ".join(fargs), formatted_args))
```

Comment:
not worrying about `self` led to a simplification, only have to handle two cases:`partial` functions, and everything else.